autodist/ray/backend.py

```diff
@@ -0,0 +1,241 @@
+                os.environ[var] = "True" if val else "False"
+            else:
+                os.environ[var] = val
+
+        # Set Ray backend to True
+        os.environ[ENV.AUTODIST_RAY_BACKEND.name] = "True"
+
+        # We either pass a strategy_builder or directly a strategy
+        self._autodist = AutoDist(strategy_builder=strategy_builder,
+                                  strategy=strategy,
+                                  resource_spec=resource_spec)
+        self._g = v1.Graph()
+        with self._g.as_default(), self._autodist.scope():
+            # model_fn and input_fn can return multiple things, pack and
+            # unpack them into the step function
+            models = model_fn()
+            inputs = input_fn()
+            if isinstance(inputs, tuple):
+                iterators = (i.get_next() if hasattr(i, 'get_next')
+                             else i for i in inputs)
+            else:
+                iterators = (inputs.get_next() if hasattr(inputs, 'get_next')
+                             else inputs,)
+
+            if not isinstance(models, tuple):
+                models = (models,)
+
+            # Create saver before creating the session
+            self._saver = autodist_saver()
+            self._fetches = train_step(*models, *iterators)
+            self._session = self._autodist.create_distributed_session()
+
+    def step(self):
+        """Take one training step."""
+        self._epoch += 1
+        return self._session.run(self._fetches)
+
+    def get_strategy(self):
+        """Fetch the current strategy."""
+        return self._autodist._strategy
+
+    def save(self, checkpoint_dir, checkpoint_prefix=""):
+        """Save a TF checkpoint."""
+        self._saver.save(self._session, checkpoint_dir + checkpoint_prefix, global_step=self._epoch + 1)
+        self._saver.restore(self._session, tf.train.latest_checkpoint(checkpoint_dir))
+
+    def restore(self, checkpoint_dir):
+        """Restore the checkpoint from the directory."""
+        with self._g.as_default(), self._autodist.scope():
+            self._saver.restore(self._session, tf.train.latest_checkpoint(checkpoint_dir))
+
+
+class TFTrainer:
+    """TFTrainer represents one training job."""
+
+    def __init__(self, strategy_builder, train_step, model_fn, input_fn):
+
+        # Set Ray backend
+        os.environ[ENV.AUTODIST_RAY_BACKEND.name] = "True"
+
+        # Go from resource_info -> ResourceSpec -> ClusterSpec
+        self._resource_spec = ResourceSpec(
+            resource_info=self._get_resource_info())
+
+        self._replicas = []   # Replica actors, also contains master
+
+        # Start TF Servers on each node of the cluster
+        self._start_tf_servers()
+
+        def spawn_replica(replica_host, strategy_builder, strategy=None, env=None):
+            # Enforce actor placement on the provided host
+            runner = ray.remote(resources={f"node:{replica_host}": 0.01},
```

Comment:
I believe this requires custom resource specification when you do `ray up` to start the ray cluster?